In [1]:
import pandas as pd
import numpy as np


In [2]:
df = pd.read_csv(r"C:\Users\aabha\Downloads\raw_data.csv", encoding = "ISO-8859-1")

In [3]:
df.head(3)

Date  Label                                               Top1  \
0  2000-01-03      0  A 'hindrance to operations': extracts from the...   
1  2000-01-04      0                                          Scorecard   
2  2000-01-05      0                  Coventry caught on counter by Flo   

                                 Top2  \
0                           Scorecard   
1                 The best lake scene   
2  United's rivals on the road to Rio   

                                            Top3  \
0                Hughes' instant hit buoys Blues   
1                  Leader: German sleaze inquiry   
2  Thatcher issues defence before trial by video   

                                            Top4  \
0       Jack gets his skates on at ice-cold Alex   
1                                  Cheerio, boyo   
2  Police help Smith lay down the law at Everton   

                                          Top5  \
0       Chaos as Maracana builds up for United   
1                     The main recommendations   
2  Tale of Trautmann bears two more retellings   

                                                Top6  \
0  Depleted Leicester prevail as Elliott spoils E...   
1                             Has Cubie killed fees?   
2                                England on the rack   

                                              Top7  \
0                 Hungry Spurs sense rich pickings   
1                           Has Cubie killed fees?   
2  Pakistan retaliate with call for video of Walsh   

                                   Top8  ...  \
0     Gunners so wide of an easy target  ...   
1                Has Cubie killed fees?  ...   
2  Cullinan continues his Cape monopoly  ...   

                                      Top16  \
0  Flintoff injury piles on woe for England   
1                      On the critical list   
2               South Melbourne (Australia)   

                                               Top17  \
0  Hunters threaten Jospin with new battle of the...   
1                          The timing of their lives   
2                                    Necaxa (Mexico)   

                                 Top18  \
0  Kohl's successor drawn into scandal   
1                          Dear doctor   
2                  Real Madrid (Spain)   

                                               Top19  \
0               The difference between men and women   
1  Irish court halts IRA man's extradition to Nor...   
2                          Raja Casablanca (Morocco)   

                                               Top20  \
0                Sara Denver, nurse turned solicitor   
1  Burundi peace initiative fades after rebels re...   
2                               Corinthians (Brazil)   

                                            Top21  \
0  Diana's landmine crusade put Tories in a panic   
1            PE points the way forward to the ECB   
2                              Tony's pet project   

                                               Top22              Top23  \
0  Yeltsin's resignation caught opposition flat-f...   Russian roulette   
1  Campaigners keep up pressure on Nazi war crime...     Jane Ratcliffe   
2                            Al Nassr (Saudi Arabia)  Ideal Holmes show   

                                               Top24  \
0                                           Sold out   
1  Yet more things you wouldn't know without the ...   
2               Pinochet leaves hospital after tests   

                          Top25  
0            Recovering a title  
1  Millennium bug fails to bite  
2                  Useful links  

[3 rows x 27 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4101 entries, 0 to 4100
Data columns (total 27 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Date    4101 non-null   object
 1   Label   4101 non-null   int64 
 2   Top1    4101 non-null   object
 3   Top2    4101 non-null   object
 4   Top3    4101 non-null   object
 5   Top4    4101 non-null   object
 6   Top5    4101 non-null   object
 7   Top6    4101 non-null   object
 8   Top7    4101 non-null   object
 9   Top8    4101 non-null   object
 10  Top9    4101 non-null   object
 11  Top10   4101 non-null   object
 12  Top11   4101 non-null   object
 13  Top12   4101 non-null   object
 14  Top13   4101 non-null   object
 15  Top14   4101 non-null   object
 16  Top15   4101 non-null   object
 17  Top16   4101 non-null   object
 18  Top17   4101 non-null   object
 19  Top18   4101 non-null   object
 20  Top19   4101 non-null   object
 21  Top20   4101 non-null   object
 22  Top21   4101 non-null   

In [5]:
train = df[df['Date'] < '20150101']
test = df[df['Date'] > '20141231']

In [6]:
# Removing Punctuation

data=train.iloc[:,2:27]
data.replace("[^a-zA-Z]"," ",regex=True, inplace=True)

In [7]:
# Renaming column names for ease of access

list1= [i for i in range(25)]
new_Index=[str(i) for i in list1]
data.columns= new_Index
data.head(3)

0  \
0  A  hindrance to operations   extracts from the...   
1                                          Scorecard   
2                  Coventry caught on counter by Flo   

                                    1  \
0                           Scorecard   
1                 The best lake scene   
2  United s rivals on the road to Rio   

                                               2  \
0                Hughes  instant hit buoys Blues   
1                  Leader  German sleaze inquiry   
2  Thatcher issues defence before trial by video   

                                               3  \
0       Jack gets his skates on at ice cold Alex   
1                                  Cheerio  boyo   
2  Police help Smith lay down the law at Everton   

                                             4  \
0       Chaos as Maracana builds up for United   
1                     The main recommendations   
2  Tale of Trautmann bears two more retellings   

                                                   5  \
0  Depleted Leicester prevail as Elliott spoils E...   
1                             Has Cubie killed fees    
2                                England on the rack   

                                                 6  \
0                 Hungry Spurs sense rich pickings   
1                           Has Cubie killed fees    
2  Pakistan retaliate with call for video of Walsh   

                                      7  \
0     Gunners so wide of an easy target   
1                Has Cubie killed fees    
2  Cullinan continues his Cape monopoly   

                                                   8  \
0      Derby raise a glass to Strupar s debut double   
1  Hopkins  furious  at Foster s lack of Hannibal...   
2             McGrath puts India out of their misery   

                                          9  ...  \
0  Southgate strikes  Leeds pay the penalty  ...   
1                    Has Cubie killed fees   ...   
2            Blair Witch bandwagon rolls on  ...   

                                         15  \
0  Flintoff injury piles on woe for England   
1                      On the critical list   
2               South Melbourne  Australia    

                                                  16  \
0  Hunters threaten Jospin with new battle of the...   
1                          The timing of their lives   
2                                    Necaxa  Mexico    

                                    17  \
0  Kohl s successor drawn into scandal   
1                          Dear doctor   
2                  Real Madrid  Spain    

                                                  18  \
0               The difference between men and women   
1  Irish court halts IRA man s extradition to Nor...   
2                          Raja Casablanca  Morocco    

                                                  19  \
0                Sara Denver  nurse turned solicitor   
1  Burundi peace initiative fades after rebels re...   
2                               Corinthians  Brazil    

                                               20  \
0  Diana s landmine crusade put Tories in a panic   
1            PE points the way forward to the ECB   
2                              Tony s pet project   

                                                  21                 22  \
0  Yeltsin s resignation caught opposition flat f...   Russian roulette   
1  Campaigners keep up pressure on Nazi war crime...     Jane Ratcliffe   
2                            Al Nassr  Saudi Arabia   Ideal Holmes show   

                                                  23  \
0                                           Sold out   
1  Yet more things you wouldn t know without the ...   
2               Pinochet leaves hospital after tests   

                             24  
0            Recovering a title  
1  Millennium bug fails to bite  
2                  Useful links  

[3 rows x 25 columns]

In [8]:
# Convertng headlines to lower case

for index in new_Index:
    data[index]=data[index].str.lower()
data.head(1)

0          1  \
0  a  hindrance to operations   extracts from the...  scorecard   

                                 2                                         3  \
0  hughes  instant hit buoys blues  jack gets his skates on at ice cold alex   

                                        4  \
0  chaos as maracana builds up for united   

                                                   5  \
0  depleted leicester prevail as elliott spoils e...   

                                  6                                  7  \
0  hungry spurs sense rich pickings  gunners so wide of an easy target   

                                               8  \
0  derby raise a glass to strupar s debut double   

                                          9  ...  \
0  southgate strikes  leeds pay the penalty  ...   

                                         15  \
0  flintoff injury piles on woe for england   

                                                  16  \
0  hunters threaten jospin with new battle of the...   

                                    17                                    18  \
0  kohl s successor drawn into scandal  the difference between men and women   

                                    19  \
0  sara denver  nurse turned solicitor   

                                               20  \
0  diana s landmine crusade put tories in a panic   

                                                  21                22  \
0  yeltsin s resignation caught opposition flat f...  russian roulette   

         23                  24  
0  sold out  recovering a title  

[1 rows x 25 columns]

In [9]:
' '.join(str(x) for x in data.iloc[1,0:25])

'scorecard the best lake scene leader  german sleaze inquiry cheerio  boyo the main recommendations has cubie killed fees  has cubie killed fees  has cubie killed fees  hopkins  furious  at foster s lack of hannibal appetite has cubie killed fees  a tale of two tails i say what i like and i like what i say elbows  eyes and nipples task force to assess risk of asteroid collision how i found myself at last on the critical list the timing of their lives dear doctor irish court halts ira man s extradition to northern ireland burundi peace initiative fades after rebels reject mandela as mediator pe points the way forward to the ecb campaigners keep up pressure on nazi war crimes suspect jane ratcliffe yet more things you wouldn t know without the movies millennium bug fails to bite'

In [10]:
headlines = []
for row in range(0,len(data.index)):
    headlines.append(' '.join(str(x) for x in data.iloc[row,0:25]))

# Using Random Forest

# By CountVectorizer

In [11]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.ensemble import RandomForestClassifier

In [12]:
# implement BAG OF WORDS

countvector = CountVectorizer(ngram_range=(2,2))
traindataset = countvector.fit_transform(headlines)

In [13]:
# implement RandomForest Classifier

randomclassifier = RandomForestClassifier(n_estimators = 200, criterion = 'entropy')
randomclassifier.fit(traindataset,train['Label'])

RandomForestClassifier(criterion='entropy', n_estimators=200)

In [14]:
# Predict for the Test Dataset

test_transform = []

for row in range(0,len(test.index)):
    test_transform.append(' '.join(str(x) for x in test.iloc[row,2:27]))
    
test_dataset = countvector.transform(test_transform)
predictions = randomclassifier.predict(test_dataset)

In [15]:
predictions

array([1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 1,
       1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0,
       1, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 1,
       1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1,
       0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1,
       1, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 1,
       1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
       1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 0, 1, 1, 0,
       1, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 0,
       1, 1, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0,
       1, 1, 1, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 1,
       0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 0, 0, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,

In [16]:
# Import library to check accuracy

from sklearn.metrics import classification_report,confusion_matrix,accuracy_score

In [17]:
matrix = confusion_matrix(test['Label'],predictions)
print(matrix)
score = accuracy_score(test['Label'],predictions)
print(score)
report = classification_report(test['Label'],predictions)
print(report)

[[141  45]
 [  5 187]]
0.8677248677248677
              precision    recall  f1-score   support

           0       0.97      0.76      0.85       186
           1       0.81      0.97      0.88       192

    accuracy                           0.87       378
   macro avg       0.89      0.87      0.87       378
weighted avg       0.88      0.87      0.87       378



** Through this, we are getting accuracy of around 86.77%

# By using Tfidf

In [18]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.ensemble import RandomForestClassifier

In [19]:
# implement BAG OF WORDS

Tfidfvector = TfidfVectorizer(ngram_range=(2,2))
traindataset = countvector.fit_transform(headlines)

In [20]:
# implement RandomForest Classifier

randomclassifier = RandomForestClassifier(n_estimators = 200, criterion = 'entropy')
randomclassifier.fit(traindataset,train['Label'])

RandomForestClassifier(criterion='entropy', n_estimators=200)

In [21]:
# Predict for the Test Dataset

test_transform = []

for row in range(0,len(test.index)):
    test_transform.append(' '.join(str(x) for x in test.iloc[row,2:27]))
    
test_dataset = countvector.transform(test_transform)
predictions = randomclassifier.predict(test_dataset)

In [22]:
predictions

array([1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 1,
       1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0,
       1, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 1,
       1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1,
       0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1,
       1, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 1,
       1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
       1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 0, 1, 1, 0,
       1, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 0,
       1, 1, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0,
       1, 1, 1, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 1,
       0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1,
       1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1,

In [24]:
matrix = confusion_matrix(test['Label'],predictions)
print(matrix)
score = accuracy_score(test['Label'],predictions)
print(score)
report = classification_report(test['Label'],predictions)
print(report)

[[137  49]
 [ 11 181]]
0.8412698412698413
              precision    recall  f1-score   support

           0       0.93      0.74      0.82       186
           1       0.79      0.94      0.86       192

    accuracy                           0.84       378
   macro avg       0.86      0.84      0.84       378
weighted avg       0.86      0.84      0.84       378



** Through this, we are getting accuracy of around 84.12% 

# Using Naive Bayes

In [26]:
from sklearn.naive_bayes import MultinomialNB

In [31]:
traindataset

<3975x584289 sparse matrix of type '<class 'numpy.int64'>'
	with 1075856 stored elements in Compressed Sparse Row format>

In [28]:
naive

sklearn.naive_bayes.MultinomialNB

In [32]:
naive = MultinomialNB()
naive.fit(traindataset,train['Label'])

MultinomialNB()

In [33]:
# Predict for the Test Dataset

test_transform = []

for row in range(0,len(test.index)):
    test_transform.append(' '.join(str(x) for x in test.iloc[row,2:27]))
    
test_dataset = countvector.transform(test_transform)
predictions = naive.predict(test_dataset)

In [34]:
predictions

array([1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 1,
       1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0,
       1, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 1,
       1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1,
       0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1,
       1, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 1,
       1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
       1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 0, 1, 1, 0,
       1, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 0,
       1, 1, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0,
       1, 1, 1, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 1,
       0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0,

In [35]:
matrix = confusion_matrix(test['Label'],predictions)
print(matrix)
score = accuracy_score(test['Label'],predictions)
print(score)
report = classification_report(test['Label'],predictions)
print(report)

[[138  48]
 [ 10 182]]
0.8465608465608465
              precision    recall  f1-score   support

           0       0.93      0.74      0.83       186
           1       0.79      0.95      0.86       192

    accuracy                           0.85       378
   macro avg       0.86      0.84      0.84       378
weighted avg       0.86      0.85      0.84       378



** Through this, we are getting accuracy of around 84.65% 